### PR TITLE
feat: Add functionality for setting the returned credentials

### PR DESCRIPTION
**Pull request by jimmidyson**
Does not call any provider plugin yet, placeholders are in place.

### CODE REVIEW

File: pkg/credentialprovider/shim/shim.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+	return credentialproviderv1beta1.AuthConfig{}, false
+}
+
+func isRegistryCredentialsOnly(cfg *v1alpha1.MirrorConfig) bool {
+	return cfg != nil &&
+		cfg.MirrorCredentialsStrategy == v1alpha1.MirrorCredentialsOnly
+}
+
+func globbedDomainForImage(img string) (string, error) {
```
Comment (dkoshkin) [inline pkg/credentialprovider/shim/shim.go:173]: This looks right, but some unit tests would be great 🙏

File: pkg/credentialprovider/shim/shim.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		switch p.cfg.Mirror.MirrorCredentialsStrategy {
+		case v1alpha1.MirrorCredentialsOnly:
+			// Only return mirror credentials by setting the image auth for the full image name whether it is already set or
+			// not.
+			authMap[img] = mirrorAuth
+		case v1alpha1.MirrorCredentialsLast:
+			// Set mirror credentials for globbed domain to ensure that the mirror credentials are used last (glob matches
+			// have lowest precedence).
+			//
+			// This means that the kubelet will first try the mirror credentials, which containerd will try against both the
```
[the remark this answers is not in the document]
Comment (jimmidyson) [reply pkg/credentialprovider/shim/shim.go:101]: What do you mean? You don't think that a full image match will be considered higher precedence?

File: pkg/credentialprovider/shim/shim.go (hunk 1 above)
Comment (jimmidyson) [reply pkg/credentialprovider/shim/shim.go]: Copied some stuff from upstream, moved globber to own package, and added unit tests.

File: pkg/credentialprovider/shim/shim.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimmidyson) [reply pkg/credentialprovider/shim/shim.go:90]: Right, that's why the placeholder code never adds any credentials matching for the mirror domain at all - they would never be used. It adds the mirror credentials but matching the image or the globbed domain depending on the specified mirror credential strategy.

In a pure airgapped environment the default credential strategy of `MirrorCredentialsOnly` would be used so only mirror credential are added, but matching the requested image itself, not the mirror domain.

I hope that makes sense 😅

**Review by dkoshkin — APPROVED** [review]

**Pull request merged**